### PR TITLE
Added Placement Preview

### PR DESCRIPTION
**Pull request by Commodore64x**
**Overview**
This Pull Request adds in the previewing of blocks before they are placed. It currently only supports 3.5 and will be updated to 4.0 at a later date.

**Contributors:**
Commodore64x (Main mechanics)
TheChosenEvilOne (Multiblock support)
Sonnicon (Turret support and opacity slider)

### CODE REVIEW

File: core/src/io/anuke/mindustry/input/PlaceMode.java [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import io.anuke.mindustry.ui.fragments.ToolFragment;
 import io.anuke.mindustry.world.Block;
 import io.anuke.mindustry.world.Tile;
+import io.anuke.mindustry.world.blocks.types.production.Drill;
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/input/PlaceMode.java]: Unused import.

File: core/src/io/anuke/mindustry/graphics/BlockRenderer.java [hunk 2]
```diff
@@ -269,15 +273,53 @@ void cacheChunk(int cx, int cy, boolean floor){
...
 	}
-}
+	
+	public void drawPreview(Block block, float drawx, float drawy, float rotation, float opacity) {
+		Draw.reset();
+		Draw.alpha(opacity);
+		Draw.rect(block.name(), drawx, drawy, rotation);
+	}
+	
+	public void handlePreview(Block block, float rotation, float drawx, float drawy, int tilex, int tiley) {
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/graphics/BlockRenderer.java]: You're not resetting color here, which leads to some random tinting issues.

File: core/assets/bundles/bundle.properties [hunk 3]
```diff
@@ -270,6 +270,7 @@ setting.multithread.name=Multithreading
 setting.fps.name=Show FPS
 setting.vsync.name=VSync
 setting.lasers.name=Show Power Lasers
+setting.previewopacity.name = Placing Preview Opacity
```
Comment (TheChosenEvilOne) [inline core/assets/bundles/bundle.properties]: Extra spaces around "=" should be removed

File: core/src/io/anuke/mindustry/graphics/BlockRenderer.java [hunk 4]
```diff
@@ -269,15 +273,53 @@ void cacheChunk(int cx, int cy, boolean floor){
...
+		
+		if(control.input().recipe != null && state.inventory.hasItems(control.input().recipe.requirements)
+		   && control.input().validPlace(tilex, tiley, block) && (android || control.input().cursorNear())) {
+			
+			float opacity = (float)Settings.getInt("previewopacity")/100f;
```
Comment (TheChosenEvilOne) [inline core/src/io/anuke/mindustry/graphics/BlockRenderer.java]: this should be moved to be after the multiblock check

File: core/src/io/anuke/mindustry/graphics/BlockRenderer.java [hunk 5]
```diff
@@ -269,15 +273,53 @@ void cacheChunk(int cx, int cy, boolean floor){
...
 	private void createBatch(){
 		if(cbatch != null)
 			cbatch.dispose();
 		cbatch = new CacheBatch(world.width() * world.height() * 4);
 	}
-}
+	
+	public void drawPreview(Block block, float drawx, float drawy, float rotation, float opacity) {
+		Draw.reset();
+		Draw.alpha(opacity);
```
Comment (TheChosenEvilOne) [inline core/src/io/anuke/mindustry/graphics/BlockRenderer.java]: Draw.alpha and Draw.reset both should be moved to the handlePreview function after the multiblock check

File: core/src/io/anuke/mindustry/graphics/BlockRenderer.java [hunk 6]
```diff
@@ -269,15 +273,53 @@ void cacheChunk(int cx, int cy, boolean floor){
...
+					Draw.rect("block-" + block.width + "x" + block.height, drawx, drawy);
+				} else {
+					Draw.rect("block", drawx, drawy);
+				}
+			}
+			
+			drawPreview(block, drawx, drawy, rotation, opacity);
+			
+			Tile tile = world.tile(tilex, tiley);
+			if((block instanceof Drill || block instanceof Pump) && block.isLayer(tile)) {
```
Comment (Anuken) [inline core/src/io/anuke/mindustry/graphics/BlockRenderer.java]: This is rather hacky, and will break with multiblock drills, pumps, or other new blocks that display a red slash. I don't think it's necessary.

File: core/src/io/anuke/mindustry/input/PlaceMode.java [hunk 7]
```diff
@@ -272,25 +276,28 @@ public void draw(int tilex, int tiley, int endx, int endy){
...
 				
 				int amount = 1;
 				for(int cx = 0; cx <= Math.abs(endx - tilex); cx ++){
 					for(int cy = 0; cy <= Math.abs(endy - tiley); cy ++){
-						int px = tx + cx * Mathf.sign(ex - tx), 
-						py = ty + cy * Mathf.sign(ey - ty);
+						int px = tx + cx * Mathf.sign(ex - tx),
+								py = ty + cy * Mathf.sign(ey - ty);
+						
+						renderer.getBlocks().handlePreview(control.input().recipe.result, control.input().recipe.result.rotate ? rotation * 90 : 0f, px * t + offset.x, py * t + offset.y, px, py);
```
Comment (TheChosenEvilOne) [inline core/src/io/anuke/mindustry/input/PlaceMode.java]: put Draw.color(Colors.get("placeInvalid")); here

**Comment by Anuken** [comment]
screenshot from 2018-04-15 19-51-01

Another small bug: The actual placement result is shown on top, while the preview version is off by a block. When I try moving the selection back one block, the last 2x2 preview block is not displayed at all.

**Comment by Commodore64x** [comment]
@Anuken This is intentional? It would look very ugly if you still display the preview despite no have actually selected a 2x2 space

**Comment by Anuken** [comment]
I don't think you understood what I meant. When I let go of the mouse, I get the same arrangement of weapon factories as shown _above the preview._ However, if you look closely at the preview, all the previewed blocks past the core are shifted to the right by one block. The preview you are displaying is incorrect and missing a block.

**Comment by Commodore64x** [comment]
No? there is a gap of 6 spaces in between, that is space for three weapons factories, even without the preview a block would not be placed there.

**Comment by Commodore64x** [comment]
Also noting, the fix for this could also be used for the 'red mess'

**Pull request merged**